src/Avalonia.Vulkan/VulkanDisplay.cs

```diff
@@ -0,0 +1,392 @@
+                AccessFlags.AccessTransferReadBit,
+                renderTarget.MipLevels);
+
+            var copyRegion = new ImageCopy
+            {
+                SrcOffset = new Offset3D { X = 0, Y = 0, Z = 0 },
+                DstOffset = new Offset3D { X = 0, Y = 0, Z = 0 },
+                Extent = new Extent3D((uint?)Size.Width, (uint?)Size.Height, 1),
+                SrcSubresource =
+                    new ImageSubresourceLayers
+                    {
+                        AspectMask = ImageAspectFlags.ImageAspectColorBit,
+                        BaseArrayLayer = 0,
+                        LayerCount = 1,
+                        MipLevel = 0
+                    },
+                DstSubresource = new ImageSubresourceLayers
+                {
+                    AspectMask = ImageAspectFlags.ImageAspectColorBit,
+                    BaseArrayLayer = 0,
+                    LayerCount = 1,
+                    MipLevel = 0
+                }
+            };
+
+            Device.Api.CmdCopyImage(commandBuffer, renderTarget.Image.ApiHandle.Value,
+                ImageLayout.TransferSrcOptimal,
+                _swapchainImages[_nextImage],
+                ImageLayout.TransferDstOptimal,
+                1,
+                copyRegion);
+
+            VulkanMemoryHelper.TransitionLayout(Device, commandBuffer,
+                renderTarget.Image.ApiHandle.Value, ImageLayout.TransferSrcOptimal,
+                AccessFlags.AccessTransferReadBit,
+                ImageLayout.ColorAttachmentOptimal,
+                AccessFlags.AccessNoneKhr,
+                renderTarget.MipLevels);
+
+            return;
+        }
+
+        internal unsafe void EndPresentation(VulkanCommandBufferPool.VulkanCommandBuffer commandBuffer)
+        {
+            VulkanMemoryHelper.TransitionLayout(Device, commandBuffer.ApiHandle,
+                _swapchainImages[_nextImage], ImageLayout.TransferDstOptimal,
+                AccessFlags.AccessNoneKhr,
+                ImageLayout.PresentSrcKhr,
+                AccessFlags.AccessNoneKhr,
+                1);
+
+            commandBuffer.Submit(
+                new[] { _semaphorePair.ImageAvailableSemaphore },
+                new[] { PipelineStageFlags.PipelineStageColorAttachmentOutputBit },
+                new[] { _semaphorePair.RenderFinishedSemaphore });
+
+            var semaphore = _semaphorePair.RenderFinishedSemaphore;
+            var swapchain = _swapchain;
+            var nextImage = _nextImage;
+
+            Result result;
+
+            var presentInfo = new PresentInfoKHR
+            {
+                SType = StructureType.PresentInfoKhr,
+                WaitSemaphoreCount = 1,
+                PWaitSemaphores = &semaphore,
+                SwapchainCount = 1,
+                PSwapchains = &swapchain,
+                PImageIndices = &nextImage,
+                PResults = &result
+            };
+
+            lock (Device.Queues)
```

Comment:
I think this needs to lock underlying `_lock` in device. Since most GPU vendors have unified queues then presentation queue is the same as graphics/compute so this is accessing it in an unsafe way,